document-policy-explainer.md

```diff
@@ -0,0 +1,341 @@
+
+```http
+Document-Policy: image-compression;bpp=1.5,no-document-write,no-unsized-media
+```
+
+While either of these headers would cause a network error to be returned
+instead:
+
+```http
+Document-Policy: no-unsized-media,image-compression;bpp=3
+Document-Policy: image-compression;bpp=2
+```
+
+If the document is successfully loaded, any subsequent navigations from that
+frame will also have their requests sent with the same
+`Sec-Required-Document-Policy` header.
+
+### Deeper nesting
+
+As above, if the top-level document uses the policy attribute in an iframe
+element, like this:
+
+```html
+<iframe src="https://img.example.com/" policy="image-compression;bpp=2">
+```
+
+Then the request for the frame will look like this:
+
+```http
+GET / HTTP/1.1
+Host: img.example.com
+Sec-Required-Document-Policy: image-compression;bpp=2
+```
+
+If the framed document then includes its own subframes, they will be sent with
+the same `Sec-Required-Document-Policy` header, as the policy set by the
+top-level document applies to all nested iframes within.
+
+If the framed document then also applies a policy attribute to its iframes, then
+the required policy indicated by those frames will combine with the required
+policy imposed on the document itself.
+
+A document like this:
+
+```html
+<html>
+  <body>
+    <iframe src="https://a.example.com/" policy="no-unsized-media"></iframe>
+    <iframe src="https://b.example.com/"
+            policy="image-compression;bpp=1"></iframe>
+    <iframe src="https://c.example.com/"
+            policy="image-compression;bpp=4"></iframe>
+  </body>
+</html>
+```
+
+Would result in these three (simplified) HTTP requests:
+
+```http
+GET / HTTP/1.1
+Host: a.example.com
+Sec-Required-Document-Policy: image-compression;bpp=2,no-unsized-media
+```
+
+```http
+GET / HTTP/1.1
+Host: b.example.com
+Sec-Required-Document-Policy: image-compression;bpp=1
+```
+
+```http
+GET / HTTP/1.1
+Host: c.example.com
+Sec-Required-Document-Policy: image-compression;bpp=2
+```
+
+(Note that in the last example, the stricter requirements imposed by the
+top-level document subsume the requirements on the nested frame, so the combined
+threshold value is still 'bpp=2'.)
+
+### Sandboxing nested content (Traditional sandbox)
+
+As examples of different ways to sandbox content using a combination of the
+`sandbox` and `policy` attributes, the following could all be used to create a
+sandboxed iframe:
+
+```html
+<iframe sandbox>
+<iframe sandbox="allow-scripts allow-forms">
+<iframe policy="no-scripts"> (This is a non-sandboxed frame with no scripting allowed)
+<iframe sandbox policy="scripts"> (Should this be allowed?)
+<iframe sandbox="allow-scripts" policy="no-scripts"> (This one needs some work)
+<iframe policy="no-same-origin"> (This is a frame with an opaque origin, but no other sandboxing)
+```
+
+None of these result in a `Sec-Required-Feature-Policy` header, as all of the
+existing sandbox features are safelisted.
```

Comment:
I think we should leave this as an open issue as surfacing existing sandboxing features seems quite valuable and gives servers more control over potential attacks. (In that case there would not have to be matching response header, but that seems okay, quite similar to CORS.) Also, this should be `Sec-Required-Document-Policy`.

Reply:
Thanks for catching the name :)

I was looking to maintain some kind of parity between the items which are advertised in that header and the policies which you *must* declare in order to be loaded. I wonder if we'd need some different syntax to denote policies which will be enforced, but which you don't get to refuse?

Leaving an open issue seems like the right thing here, agreed.